cupyx/scipy/sparse/linalg/_lobpcg.py

```diff
@@ -0,0 +1,662 @@
+                  in list_obj for arr in list_iter)
+    order = 'F' if F_order and not C_order else 'C'
+    result = cupy.empty(tuple(final_shape), dtype=dtype, order=order)
+
+    start_idx_row = 0
+    start_idx_col = 0
+    end_idx_row = 0
+    end_idx_col = 0
+    for i in range(n_rows):
+        end_idx_row = start_idx_row + list_obj[i][0].shape[0]
+        start_idx_col = 0
+        for j in range(n_cols):
+            end_idx_col = start_idx_col + list_obj[i][j].shape[1]
+            result[start_idx_row:end_idx_row,
+                   start_idx_col: end_idx_col] = list_obj[i][j]
+            start_idx_col = end_idx_col
+        start_idx_row = end_idx_row
+    return result
+
+
+def _report_nonhermitian(M, name):
+    """
+    Report if `M` is not a hermitian matrix given its type.
+    """
+
+    md = M - M.T.conj()
+
+    nmd = linalg.norm(md, 1)
+    tol = 10 * cupy.finfo(M.dtype).eps
+    tol *= max(1, float(linalg.norm(M, 1)))
+    if nmd > tol:
+        print('matrix %s of the type %s is not sufficiently Hermitian:'
+              % (name, M.dtype))
+        print('condition: %.e < %e' % (nmd, tol))
+
+
+def _as2d(ar):
+    """
+    If the input array is 2D return it, if it is 1D, append a dimension,
+    making it a column vector.
+    """
+    if ar.ndim == 2:
+        return ar
+    else:  # Assume 1!
+        aux = cupy.array(ar, copy=False)
+        aux.shape = (ar.shape[0], 1)
+        return aux
+
+
+def _makeOperator(operatorInput, expectedShape):
+    """Takes a dense numpy array or a sparse matrix or
+    a function and makes an operator performing matrix * blockvector
+    products.
+    """
+    if operatorInput is None:
+        return None
+    else:
+        operator = splinalg.aslinearoperator(operatorInput)
+
+    if operator.shape != expectedShape:
+        raise ValueError('operator has invalid shape')
+
+    return operator
+
+
+def _applyConstraints(blockVectorV, YBY, blockVectorBY, blockVectorY):
+    """Changes blockVectorV in place."""
+    YBV = cupy.dot(blockVectorBY.T.conj(), blockVectorV)
+    # awaiting the implementation of cho_solve in PR #4172
+    # tmp = cho_solve(factYBY, YBV)
+    tmp = linalg.solve(YBY, YBV)
+    blockVectorV -= cupy.dot(blockVectorY, tmp)
+
+
+def _b_orthonormalize(B, blockVectorV, blockVectorBV=None, retInvR=False):
+    """B-orthonormalize the given block vector using Cholesky."""
+    normalization = blockVectorV.max(axis=0)+cupy.finfo(blockVectorV.dtype).eps
+    blockVectorV = blockVectorV / normalization
+    if blockVectorBV is None:
+        if B is not None:
+            blockVectorBV = B(blockVectorV)
+        else:
+            blockVectorBV = blockVectorV
+    else:
+        blockVectorBV = blockVectorBV / normalization
+    VBV = cupy.matmul(blockVectorV.T.conj(), blockVectorBV)
+    try:
+        # VBV is a Cholesky factor
+        VBV = linalg.cholesky(VBV).T
+        VBV = cupy.asarray(VBV)
+        assert not isinstance(VBV, type(None))
```

Comment:
This assertion is not needed 😃

Reply:
An assertion is remaining.: https://github.com/cupy/cupy/pull/4281/files#diff-b39ae3a6aae7508d9d618fa66b1defb211617ec2b3440f8f603025217751ba6aR331